stream.go

```diff
@@ -264,24 +242,23 @@ func (s *Stream) Reset() error {
 	return nil
 }
 
+func (s *Stream) cancelDeadlines() {
+	s.rDeadline.set(time.Time{})
+	s.wDeadline.set(time.Time{})
+}
+
 func (s *Stream) SetDeadline(t time.Time) error {
-	s.deadlineLock.Lock()
-	defer s.deadlineLock.Unlock()
-	s.rDeadline = t
-	s.wDeadline = t
+	s.rDeadline.set(t)
```

Comment:
We should try to prevent this from happening after closing the stream so we don't leak a timer. Not that much of an issue but is there some kind of state lock we can take?

Reply:
We could try the `clLock` and check if it has been closed.

Reply:
SGTM.

Reply:
I used the `clLock` and return an error if the stream has been closed.

Reply:
LGTM.